packages/core/src/common/message-service-protocol.ts

```diff
@@ -16,25 +16,64 @@
 
 import { injectable, inject } from 'inversify';
 import { ILogger } from './logger';
+import { CancellationToken } from './cancellation';
 
 export const messageServicePath = '/services/messageService';
 
 export enum MessageType {
     Error = 1,
     Warning = 2,
     Info = 3,
-    Log = 4
+    Log = 4,
+    Progress = 5
 }
 
 export interface Message {
-    type: MessageType;
-    text: string;
-    actions?: string[];
-    options?: MessageOptions;
+    readonly type?: MessageType;
+    readonly text: string;
+    readonly actions?: string[];
+    readonly options?: MessageOptions;
+}
+
+export interface ProgressMessage extends Message {
+    readonly type?: MessageType.Progress;
+    readonly options?: ProgressMessageOptions;
+}
+export namespace ProgressMessage {
+    export const Cancel = 'Cancel';
+    export function isCancelable(message: ProgressMessage): boolean {
+        return !message.options
+            || message.options.cancelable === undefined
+            || message.options.cancelable === true;
+    }
 }
 
 export interface MessageOptions {
-    timeout?: number;
+    readonly timeout?: number;
+}
+
+export interface ProgressMessageOptions extends MessageOptions {
+    /**
+     * Default: `true`
+     */
+    readonly cancelable?: boolean;
+}
+
+export interface Progress {
+    readonly id: string;
+    readonly report: (update: ProgressUpdate) => void;
+    readonly cancel: () => void;
+    readonly result: Promise<string | undefined>;
+}
+
+export interface ProgressUpdate {
+    readonly message?: string;
+    readonly progressRate?: ProgressRate;
```

Comment:
1. IMO "rate" suggest a percentage to be shown, which I'd like to avoid in the API. it's up to the renderer to decide about the presentation.
2. why not just a `work?: { done: number, total: number }`? does that type provide any value?

Reply:
fixed